doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,106 @@
+read, the major work will be subclassing :class:`~ray.data.Datasource.Reader` for 
+MongoDB. What it does is creating a list of :class:`~ray.data.ReadTask` for the given 
+list of MongDB queries. Each :class:`~ray.data.ReadTask` will return a list of blocks when called, and 
+the :class:`~ray.data.ReadTask` are executed in remote workers to parallelize the execution.
+
+First of all, let's handle a single MongDB query, as this is the execution unit in 
+:class:`~ray.data.ReadTask`. We need to connect to MongDB, execute the query against it, 
+and then convert results into Arrow format. The ``PyMongo`` and  ``PyMongoArrow`` are 
+used to achieve these.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __read_single_query_start__
+    :end-before: __read_single_query_end__
+
+Once we have this building block, we can just apply it for each provided MongoDB 
+queries and get the implementation of :class:`~ray.data.Datasource.Reader`.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __mongo_datasource_reader_start__
+    :end-before: __mongo_datasource_reader_end__
+
+-------------
+Write support
+-------------
+
+Similar to read support, we start with handling a single block. Again 
+the ``PyMongo`` and  ``PyMongoArrow`` are used for MongoDB interactions.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __write_single_block_start__
+    :end-before: __write_single_block_end__
+
+To write multiple blocks in parallel, we again use Ray remote workers
+to launch them in parallel.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __write_multiple_blocks_start__
+    :end-before: __write_multiple_blocks_end__
+
+------------
+Put together
+------------
+
+With ``_MongoDatasourceReader`` and ``_write_multiple_blocks`` above, we are 
+ready to implement :meth:`create_reader() <ray.data.Datasource.create_reader>` 
+and :meth:`do_write() <ray.data.Datasource.do_write>`, and put together 
+into a ``MongoDatasource``.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __mongo_datasource_start__
+    :end-before: __mongo_datasource_end__
+
+Now you can create a Ray Dataset from and write back to MongoDB, just like 
+any other data sources!
+
+.. code-block:: python
+
+    # Read from MongoDB datasource.
+    # The args are passed to MongoDatasource.create_reader().
+    ds = ray.data.read_datasource(
+        MongoDatasource(),
+        uri=MY_URI, database=MY_DATABASE,
+        collection=MY_COLLECTION, pipelines=MY_QUERIES
+    )
+
+    # Data processing with Dataset APIs
+    # ....
+
+    # Write to MongoDB datasource.
+    # The args are passed to MongoDatasource.do_write().
+    ds.write_datasource(
+        MongoDatasource(), uri=MY_URI, database=MY_DATABASE, collection=MY_COLLECTION
+    )
```

Comment:
Could we move this to the `doc_code/custom_datasources.py` as well?

Reply:
This part is not runnable, as it's lacking a hosted public MongoDB (like S3).